data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeEvaluatorListener.java

```diff
@@ -0,0 +1,114 @@
+
+    private final OperatorProvider operatorProvider;
+    private final ParseTreeCoercionService coercionService;
+    private final Stack<Integer> operatorSymbolStack;
+    private final Stack<Object> operandStack;
+    private final Event event;
+
+    public ParseTreeEvaluatorListener(final OperatorProvider operatorProvider,
+                                      final ParseTreeCoercionService coercionService,
+                                      final Event event) {
+        this.coercionService = coercionService;
+        this.operatorProvider = operatorProvider;
+        this.event = event;
+        operatorSymbolStack = new Stack<>();
+        operandStack = new Stack<>();
+    }
+
+    public Object getResult() {
+        if (operandStack.size() != 1) {
+            throw new IllegalStateException("The ParseTreeEvaluatorListener has not been walked through exactly once by " +
+                    "a ParseTreeWalker.");
+        }
+        return operandStack.peek();
+    }
+
+    @Override
+    public void visitTerminal(TerminalNode node) {
+        final int nodeType = node.getSymbol().getType();
+        if (nodeType == DataPrepperExpressionParser.EOF) {
+            return;
+        }
+        if (operatorProvider.containsOperator(nodeType) || nodeType == DataPrepperExpressionParser.LPAREN) {
+            operatorSymbolStack.push(nodeType);
+        } else if (nodeType == DataPrepperExpressionParser.RPAREN) {
+            // pop LPAREN at operatorSymbolStack top
+            operatorSymbolStack.pop();
+        } else {
+            final Object arg = coercionService.coercePrimaryTerminalNode(node, event);
+            operandStack.push(arg);
+        }
+    }
+
+    @Override
+    public void visitErrorNode(ErrorNode node) {
+        throw new RuntimeException("Hit error node in the parse tree: " + node.getText());
+    }
+
+    @Override
+    public void exitEveryRule(ParserRuleContext ctx) {
+        if (!operatorSymbolStack.isEmpty()) {
+            final int operatorSymbol = operatorSymbolStack.peek();
+            if (operatorSymbol != DataPrepperExpressionParser.LPAREN) {
+                final Operator<?> op = operatorProvider.getOperator(operatorSymbol);
+                if (op.shouldEvaluate(ctx)) {
+                    operatorSymbolStack.pop();
+                    try {
+                        performSingleOperation(op);
+                    } catch (final Exception e) {
+                        throw new ExpressionEvaluationException("Unable to evaluate the part of input statement: "
+                                + getPartialStatementFromContext(ctx), e);
+                    }
+                }
+            }
+        }
+    }
+
+    private void performSingleOperation(final Operator<?> operator) {
+        final int numOfArgs = operator.getNumberOfOperands();
+        final Object[] args = new Object[numOfArgs];
+        for (int i = numOfArgs - 1; i >= 0; i--) {
```

Comment:
Checking `operandStack.size() >= numOfArgs` and throwing a more informative error message may lead to a better customer experience.